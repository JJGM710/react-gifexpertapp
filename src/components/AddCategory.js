import React, { useState } from 'react'
import PropTypes from 'prop-types';

const AddCategory = ({setCategories, categories}) => {

    const [inputValue,setInputValue]= useState('')

    const handleInputChange = (e) => {
        //console.log(e.target.value);
        setInputValue(e.target.value)


        console.log('llamdo de evento onchange')
    }

    const handleSubmit = (e) =>{
        e.preventDefault();
        //console.log('handleSubmit', inputValue)
        let test = categories.find((ctg) => ctg.toLowerCase() === inputValue.trim().toLowerCase());
        if (inputValue.trim().length > 2 && test === undefined)
        {
            setCategories(ctg => [inputValue, ...ctg]);
            console.log('se hizo el cambio')
            setInputValue('');
        }
       // console.log(categories);
        
    }
    return (
        <form onSubmit={handleSubmit}>
            <p> {inputValue} </p>
            <input 
                placeholder="Search.."
                className='box'
                type = "text" 
                value = {inputValue}
                onChange = {handleInputChange}
            ></input>
        </form>
    )
}

AddCategory.propTypes = {
    setCategories : PropTypes.func.isRequired,
    categories: PropTypes.array.isRequired
};

export default AddCategory;
