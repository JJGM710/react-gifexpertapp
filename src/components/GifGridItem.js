import React from 'react'
import { PropTypes } from 'prop-types';

export const GifGridItem = ({id, title,url}) => {
   // console.log(id, title, url);
    return (
        <div className="card animate__animated animate__fadeIn animate__delay-1s">
            <img src={url} alt={title} />
            <p> {title?title.toLowerCase():'titulo no definido'} </p>
        </div>
    )
}


//definicon de default props y de isrequired


//validacion de propiedades
GifGridItem.propTypes = {
    title : PropTypes.string.isRequired,
    url : PropTypes.string.isRequired
};

//definir propiedades por defect
GifGridItem.defaultProps = {
};


// 1.enzyme
// 2.enzyme to jSON
// 3. debe mostar el comp correctamente
    
// shallow
// wrapper 
// expect wrapper toMatchSnapshot 
