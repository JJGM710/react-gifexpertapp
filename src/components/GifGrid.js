import React from 'react'
import { GifGridItem } from './GifGridItem';
//import { getGifs } from './../helpers/getGifs';
import { useFetchGifs } from './../hooks/useFetchGifs';

export const GifGrid = ({category}) => {

    // const [images, setimages] = useState([]);

    // // genera una dependecia con category, es decir, que cada vez que el valor de esa propiedad cambie va a renderizar ese component
    // useEffect(() => {
    //     getGifs(category)
    //         .then(imgs => setimages(imgs));
    // }, [category])

    const {data : images, loading} = useFetchGifs(category);


    //getGifs();
    return (
        <>
        <h3 className="animate__animated animate__flash animate__slow">{category}</h3>

        {loading && <h2 className="animate__animated animate__flash animate__slower">Loading...</h2>}

        <div className= 'card-grid'>
            
            {images.map((img) => (
                 <GifGridItem 
                    key = {img.id}
                    {...img} />
            ))
            }

        </div>
        </>
    );
};
