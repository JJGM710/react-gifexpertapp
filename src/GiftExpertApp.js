import React, { useState } from 'react'
import AddCategory from './components/AddCategory'
import { GifGrid } from './components/GifGrid';

const GiftExpertApp = () => {

    //uso de useState para establecer categorias
    const [categories, setCategories] = useState([]);

    // const handleAdd = () => {
    //     setCategories( ctgs => [...ctgs,'Kimetsu no Yaiba']);

    // }


    return (
        <>
        <h1>{"<"}Buscador de gifs{">"}</h1>
        {/* conectar con el componente AddCategory  enviando dos propiedades para establecer mas catergorias */}
        <AddCategory setCategories={setCategories} categories={categories}></AddCategory>
        <br></br>
        <ol>
        {/* con el metodo map se la usa con el estado de categorias, la cual es un array
        y renderiza varios Gifgrid como elementos dentro de categories exista.
        Cuando se anaden nuevos, debido a como trabaja React solo se rederiza esa parte nueva del DOM
        y se anade solamente el GifGrid con el elemento que fue anadido a categories */}
            {categories.map(category => (
                <GifGrid 
                    category= {category}
                    key = {category} 
                />
            ))
            }
        </ol>

        </>
    );
};



export default GiftExpertApp
