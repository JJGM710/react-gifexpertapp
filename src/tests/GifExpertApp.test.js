import  React from 'react';
import { shallow } from 'enzyme';
import '@testing-library/jest-dom';
import GiftExpertApp from './../GiftExpertApp';
import { GifGrid } from './../components/GifGrid';

describe('Pruebas en el componente principal GifExpertApp', () => {
    test('debe hacer math con el snapshot, renderizar ok', () => {
        
        const wrapper =shallow(<GiftExpertApp/>)

        expect(wrapper).toMatchSnapshot();
    })

    test('debe crear tantos GifGrid como tantos elemento hayan en categories,', () => {
        const categorias = ['Kimetsu no Yaiba', 'Dragon Ball'];

        const wrapper =shallow(<GiftExpertApp defaulCategories={categorias}/>)
        
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('GifGrid').length).toBe(categorias.length);
    })
    
    
})
