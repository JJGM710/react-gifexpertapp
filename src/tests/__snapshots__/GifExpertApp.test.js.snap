// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente principal GifExpertApp debe crear tantos GifGrid como tantos elemento hayan en categories, 1`] = `
<Fragment>
  <h1>
    &lt;
    Buscador de GIFS
    &gt;
  </h1>
  <AddCategory
    categories={
      Array [
        "Kimetsu no Yaiba",
        "Dragon Ball",
      ]
    }
    setCategories={[Function]}
  />
  <br />
  <ol>
    <GifGrid
      category="Kimetsu no Yaiba"
      key="Kimetsu no Yaiba"
    />
    <GifGrid
      category="Dragon Ball"
      key="Dragon Ball"
    />
  </ol>
</Fragment>
`;

exports[`Pruebas en el componente principal GifExpertApp debe hacer math con el snapshot, renderizar ok 1`] = `
<Fragment>
  <h1>
    &lt;
    Buscador de GIFS
    &gt;
  </h1>
  <AddCategory
    categories={Array []}
    setCategories={[Function]}
  />
  <br />
  <ol />
</Fragment>
`;
